File: Python/oomph-data-ws/uf-309-lib/ABL-C.py
# https://atcoder.jp/contests/abl/tasks/abl_c

from unionfind306 import UnionFind

n, m = map(int, input().split())

count = n + 1
uf = UnionFind(count)

for i in range(m):
    a, b = map(int, input().split())
    if uf.union(a, b):
        count -= 1

print(count - 2)

File: CSharp/Oomph.Data/UF09Test/UFs/v301/Static1/ABL_C.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Oomph.Data.UF09Lib.UFs.v301;

namespace UF09Test.UFs.v301.Static1
{
	// Test: https://atcoder.jp/contests/abl/tasks/abl_c
	// Test: https://atcoder.jp/contests/arc032/tasks/arc032_2
	class ABL_C
	{
		static int[] Read() => Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
		static (int, int) Read2() { var a = Read(); return (a[0], a[1]); }
		static void Main() => Console.WriteLine(Solve());
		static object Solve()
		{
			var (n, m) = Read2();
			var es = Array.ConvertAll(new bool[m], _ => Read2());

			var uf = new UnionFind(n + 1);
			foreach (var (a, b) in es)
				uf.Union(a, b);
			return uf.GroupsCount - 2;
		}
	}
}
